### ISSUE TITLE
dev-master - CorruptComponentPayloadException with nested components and out of order keys?

### ISSUE THREAD

**Issue by jessarcher** [livewire/livewire#1474]
### Description

I've come across a bit of an unusual scenario that causes the following error with `dev-master` but not `1.3`:
```
Livewire\Exceptions\CorruptComponentPayloadException
Livewire encountered corrupt data when trying to hydrate the [diary] component. Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests.
```

### Steps to reproduce

This is about simplest and most generic scenario I could come up with to replicate it (note it happens with Eloquent models as well as standard objects).

1. Create the following two components "Diary" and "DiaryEntry":

   ```php
   # /App/Http/Livewire/Diary.php
   <?php

   namespace App\Http\Livewire;

   use Livewire\Component;

   class Diary extends Component
   {
       public $newDiaryEntryContent = '';

       public function render()
       {
           return view('livewire.diary', [
               'diaryEntries' => [
                   '2020-09-06' => (object) [
                       'id' => 2,
                       'content' => 'Is it me you\'re looking for',
                   ],
                   '2020-09-03' => (object) [
                       'id' => 1,
                       'content' => 'Hello',
                   ],
               ]
           ]);
       }
   }
   ```
   ```php
   # /resources/views/livewire/diary.blade.php
   <div>
       @foreach ($diaryEntries as $diaryEntry)
           @livewire('diary-entry', ['diaryEntry' => $diaryEntry], key($diaryEntry->id))
       @endforeach
       <input type="text" wire:model="newDiaryEntryContent" />
   </div>
   ```
   ```php
   # /App/Http/Livewire/DiaryEntry.php
   <?php

   namespace App\Http\Livewire;

   use Livewire\Component;

   class DiaryEntry extends Component
   {
       public $content;

       public function mount($diaryEntry)
       {
           $this->content = $diaryEntry->content;
       }

       public function render()
       {
           return view('livewire.diary-entry');
       }
   }
   ```
   ```php
   # /resources/views/livewire/diary.blade.php
   <div>
       {{ $content }}
   </div>
   ```

2. Add the Diary componet to a page (E.g. the `welcome.blade.php` template that comes with Laravel)
   ```
   @livewire('diary')
   ```

3. Type in the text input and get this error:
   ```
   Livewire\Exceptions\CorruptComponentPayloadException
   Livewire encountered corrupt data when trying to hydrate the [diary] component. Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests.
   ```

**What is interesting is that any of the following things "fixes" the problem:**

1. Reorder the "diaryEntries" array so that the keys are in alphabetical order. i.e.
   ```php
               'diaryEntries' => [
                   '2020-09-03' => (object) [
                       'id' => 1,
                       'content' => 'Hello',
                   ],
                   '2020-09-06' => (object) [
                       'id' => 2,
                       'content' => 'Is it me you\'re looking for',
                   ],
               ]
   ```
2. Change the nested component `key` to something other than the `id`. E.g.
   ```php
           @livewire('diary-entry', ['diaryEntry' => $diaryEntry], key($diaryEntry->content))
   ```
3. Don't use a nested component - just add everything to the main loop.

It *seems* similar to the checksum issue with key ordering mentioned in the troubleshooting guide, but it's weird that it works fine in 1.3.

**Context**
- Livewire version: dev-master d339e82fe10272ff1ef7b1ae0d77bcf93671dba4
- Laravel version: 7.27.0 (also with 8.x beta)
- Browser: Firefox

**Comment by calebporzio** [comment]
Ah, thanks for submitting this. I'll be looking into this shortly.

**Comment by calebporzio** [comment]
@jessarcher, Thank you for providing such a fantastic issue. Clearly you boiled a vague problem down to it's simplest form - saved me so much time and overhead.

I've identified the issue and fixed it.

It's related to JavaScript re-ordering indexes because of `JSON.parse` - which then throws a checksum error because the array has changed since it left the server.

The reason this didn't happen in V1, was that we only checksumed the component property data and the component name and id. In V2 I checksum much more of the payload.

Anyhow, should be fixed on the master branch. Re-open if it's not.

Thanks again!
